File: dataStructure/树/array_binary_tree.ts
type Order = 'pre' | 'in' | 'post';

class ArrayBinaryTree {
  #tree: (number | null)[];

  constructor(arr: (number | null)[]) {
    this.#tree = arr;
  }

  // 列表容量
  size(): number {
    return this.#tree.length;
  }

  // 获取索引为 i 节点的值
  val(i: number): number | null {
    if(i<0 || i > this.size()) return null;
    return this.#tree[i];
  }

  // 获取索引为i节点的左子节点的索引
  
  left(i: number): number {
    return 2 * i + 1;
  }

  // 获取索引为i节点的右子节点的索引
  right(i: number): number {
    return 2 * i + 2;
  }

  // 获取索引为i节点的父节点的索引
  parent(i: number): number {
    return Math.floor((i-1)/2);
  }

  // 层序遍历
  levelOrder(): number[] {
    let res: number[] = [];
    for(let i=0;i<this.size();i++) {
      if(this.val(i) !== null) res.push(this.val(i) as number);
    }
    return res;
  }

  // 深度优先遍历
  #dfs(i: number, order: Order, res: (number | null)[]): void {
    // 若为空位，则返回
    if(this.val(i) === null) return;
    if(order === 'pre') res.push(this.val(i));
    this.#dfs(this.left(i), order, res);
    if(order === 'in') res.push(this.val(i));
    this.#dfs(this.right(i), order, res);
    if(order === 'post') res.push(this.val(i));
  }

  // 前序遍历
  preOrder(): (number | null)[] {
    const res: (number | null)[] = [];
    this.#dfs(0, 'pre', res);
    return res;
  }

  // 中序遍历
  inOrder(): (number | null)[] {
    const res: (number | null)[] = [];
    this.#dfs(0, 'in', res);
    return res;
  } 

  // 后序遍历
  postOrder(): (number | null)[] {
    const res: (number | null)[] = [];
    this.#dfs(0, 'post', res);
    return res;
  }
}

export default ArrayBinaryTree;

File: dataStructure/树/ArrayBinaryTree.java
package dataStructure.树;

import java.util.ArrayList;
import java.util.List;

// 数组表示下的二叉树类
public class ArrayBinaryTree {
    private List<Integer> tree;

    public ArrayBinaryTree(List<Integer> arr) {
        tree = new ArrayList<>(arr);
    }

    // 列表容量
    public int size() {
        return tree.size();
    }

    // 获取索引为i节点的值
    public Integer val(int i) {
        if (i<0 || i>= size()) return null;
        return tree.get(i);
    }

    // 获取索引为i节点的左子节点的索引
    public Integer left(int i) {
        return 2 * i + 1;
    }

    // 获取索引为i节点的右子节点的索引
    public Integer right(int i) {
        return 2 * i + 2;
    }

    // 获取索引为i节点的父节点的索引
    public Integer parent(int i) {
        return (i - 1) / 2;
    }

    // 层序遍历(因为是数组表示的，直接遍历数组就行了)
    public List<Integer> levelOrder() {
        List<Integer> res = new ArrayList<>();
        for (int i=0;i<size();i++) {
            if (val(i) != null) res.add(val(i));
        }
        return res;
    }

    // 深度优先遍历
    private void dfs(Integer i, String order, List<Integer> res) {
        // 如果是空位，则返回
        if (val(i) == null) return;
        // 前序遍历
        if ("pre".equals(order)) res.add(val(i));
        dfs(left(i), order, res);
        // 中序遍历
        if ("in".equals(order)) res.add(val(i));
        dfs(right(i), order, res);
        // 后序遍历
        if ("post".equals(order)) res.add(val(i));
    }

    // 前序遍历
    public List<Integer> preOrder() {
        List<Integer> res = new ArrayList<>();
        dfs(0, "pre", res);
        return res;
    }

    // 中序遍历
    public List<Integer> inOrder() {
        List<Integer> res = new ArrayList<>();
        dfs(0, "in", res);
        return res;
    }

    // 后序遍历
    public List<Integer> postOrder() {
        List<Integer> res = new ArrayList<>();
        dfs(0, "post", res);
        return res;
    }
}
